rework storage control decision making it a bit more flexible

tomeichlersmith opened the pull request:
Separating the 'must' keywords into their own tier helps give them greater importance that reflects their name. In order, the decision follows the criteria below only using hints that match one of the listening rules configured for the process.

1. If any hint is mustDrop, drop the event.
2. If any hint is mustKeep (without any mustDrop), keep the event.
3. Simple majority decision of votes made for keep and drop.
4. The default decision is made in the event of a tie (or no votes).

This resolves #47 by enabling an "OR" system whereby different processors can use `mustKeep` for events they desire to keep no matter what the other processors decide and `NoOpinion` or `should*` hints when a processor is alright with leaving the decision to other processors in the sequence.

tomeichlersmith committed:
rework storage control decision making to be a bit more flexible

Separating the 'must' keywords into their own tier helps give them
greater importance that reflects their name. In order, the decision
follows the criteria below only using hints that match one of the
listening rules configured for the process.

1. If any hint is mustDrop, drop the event.
2. If any hint is mustKeep (without any mustDrop), keep the event.
3. Simple majority decision of votes made for keep and drop.
4. The default decision is made in the event of a tie (or no votes).

bryngemark reviewed (approved):
Thanks for getting back to this! Looks good to me.

tomeichlersmith commented:
Alright, with that approval, I'm just going to do some refactor/cleanup stuff to make the regex parsing a bit safer and easier to understand.

### To Do
- [x] refactor to use [std::regex](https://en.cppreference.com/w/cpp/regex)
- [x] move enum into StorageControl to make referencing easier (a la https://github.com/LDMX-Software/fire/blob/b65921f52f4e1f669a9e6fba3d983d8de0f86412/include/fire/StorageControl.h#L23-L37)
- [x] move filtering of hints we will listen to into `addHint` (i.e. just don't add the hints that aren't going to be listened to)

tomeichlersmith committed:
refactor StorageControl to use std::regex

This use std::regex to help make the regex parsing/matching code a bit
more robust and readable. In addition, this refactor does a few other
updates

- use enum class within StorageControl rather than a typedef outside
- put the old names for the storage control hints as constexpr for
  backwards compatibilitiy
- move the filtering of hints based on the listening rules to when they
  are added, a hint isn't cached in the hints_ member unless it passes
one of the listening rules
- remove the Rule and Hint struct since they are no longer needed with
  this design

tomeichlersmith reviewed (commented):
Alright, the refactor is all done, this is ready to merge.
